### PR TITLE
mostly cleanup which helper

### PR DESCRIPTION
**Pull request by lamont-granquist**
trying to move towards reducing code duplication with chef/chef

### CODE REVIEW

File: lib/mixlib/shellout/windows.rb [hunk 1]
```diff
@@ -289,35 +288,17 @@ def should_run_under_cmd?(command)
         return false
       end
 
-      def pathext
-        @pathext ||= ENV["PATHEXT"] ? ENV["PATHEXT"].split(";") + [""] : [""]
-      end
-
-      # which() mimicks the Unix which command
-      # FIXME: it is not working
```
Comment (thommay) [inline lib/mixlib/shellout/windows.rb]: that fixme is amazeballs

Comment (lamont-granquist) [inline lib/mixlib/shellout/windows.rb]: yeah, its totes fun to try to figure out why.

i don't doubt it isn't working, but its a mess and hosh's tests are very stubby and fairly clever and hard to discern how all the pieces fit together.

**Comment by lamont-granquist** [comment]
Okay this is a minor refactor of the `which` logic to make it slightly less awful

The bigger thing here is the major refactor of the spec tests which remove some of the hugely stubby tests and do a better test of the more public-ish `#command_to_run` API.  I think I've preserved all the existing use cases.  The specs are a bit easier to understand what the inputs and outputs are now.  Validated the specs now pass on the old code and the new code.

File: lib/mixlib/shellout/windows.rb [hunk 2]
```diff
@@ -289,35 +276,26 @@ def should_run_under_cmd?(command)
...
+              return filename if File.executable?(filename) && !File.directory?(filename)
+            end
+          end
         end
-        return nil
-      end
-
-      def executable?(path)
-        File.executable?(path) && !File.directory?(path)
+        #end
```
Comment (coderanger) [inline lib/mixlib/shellout/windows.rb]: Vestigial :)

File: lib/mixlib/shellout.rb [hunk 3]
```diff
@@ -253,7 +253,7 @@ def exitstatus
     #   within +timeout+ seconds (default: 600s)
     def run_command
       if logger
-        log_message = (log_tag.nil? ? "" : "#@log_tag ") << "sh(#@command)"
+        log_message = (log_tag.nil? ? "" : "#{@log_tag} ") << "sh(#{@command})"
```
Comment (coderanger) [inline lib/mixlib/shellout.rb]: :lolwut:

Comment (lamont-granquist) [inline lib/mixlib/shellout.rb]: yep

**Comment by lamont-granquist** [comment]
@coderanger i did a slight API change to throw a new `Mixlib::ShellOut::EmptyWindowsCommand` error because the old behavior threw some obtuse CreateProcessW argument error that was unhelpful.  I don't /think/ you'd care about that, but stranger things have happened.  Poise tests are still green against this.

**Comment by lamont-granquist** [comment]
@chef/client-core squashing, ready for any final review

**Comment by coderanger** [comment]
@lamont-granquist Still :+1: from me

**Pull request merged**